### PR TITLE
Fix Race in jar upload during hadoop indexing

### PR DESCRIPTION
**Pull request by nishantmonu51**
fixes Race while uploading jar files. 
- SNAPSHOT jar files are always uploaded to job specific non-shared directory and cleaned up with the job cleanup. 
- Non-Snapshot jar files are first uploaded to temporary intermediate path and then renamed to the shared directory without overwriting any existing file. this fixes the race when multiple jobs sees the jar file as not present and try to upload it simultaneously. 
 
Fixes - https://github.com/druid-io/druid/issues/582

### CODE REVIEW

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 1]
```diff
@@ -109,34 +121,92 @@ public static void setupClasspath(
...
+    // Non-snapshot jar files are uploaded to the shared classpath.
+    final Path hdfsPath = new Path(distributedClassPath, jarFile.getName());
+    if (!fs.exists(hdfsPath)) {
+      // Muliple jobs can try to upload the jar here,
+      // to avoid them from overwriting files, first upload to intermediateClassPath and then rename to the distributedClasspath.
+      final Path intermediateHdfsPath = new Path(intermediateClassPath, jarFile.getName());
+      uploadJar(jarFile, intermediateHdfsPath, fs);
+      try {
+        log.info("Renaming jar to path[%s]", hdfsPath);
+        fs.rename(intermediateHdfsPath, hdfsPath);
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: If this races can it leave intermediateHdfsPath stale in the distributed FS?

Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: the intermediate path is in job working directory which gets cleaned up during job cleanup. 
Also, fixed it to clean the intermediate jar file here.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 2]
```diff
@@ -109,34 +121,96 @@ public static void setupClasspath(
...
+        }
+      }
+      catch (IOException e) {
+        // rename failed, possibly due to race condition. check if some other job has uploaded the jar file.
+        if (!fs.exists(hdfsPath)) {
+          log.error(e, "IOException while Renaming jar file");
+          throw e;
+        }
+      } finally {
+        if (fs.exists(intermediateHdfsPath)) {
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Exists and delete can throw IOException, you'll want to catch those and at least log them. This is a bit tricky

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 3]
```diff
@@ -109,34 +121,96 @@ public static void setupClasspath(
...
+      try {
+        log.info("Renaming jar to path[%s]", hdfsPath);
+        fs.rename(intermediateHdfsPath, hdfsPath);
+        if(!fs.exists(hdfsPath)){
+          throw new ISE("File does not exist even after moving from[%s] to [%s]", intermediateHdfsPath, hdfsPath);
+        }
+      }
+      catch (IOException e) {
+        // rename failed, possibly due to race condition. check if some other job has uploaded the jar file.
+        if (!fs.exists(hdfsPath)) {
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: exists can throw IOException and mask the caught exception

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 2 above)
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: ideally the exceptions would bubble up as suppressed exceptions, but that may not be possible to do in a clean way.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 4]
```diff
@@ -109,34 +121,96 @@ public static void setupClasspath(
...
+  {
+    log.info("Uploading jar to path[%s]", path);
+    ByteStreams.copy(
+        Files.newInputStreamSupplier(jarFile),
+        new OutputSupplier<OutputStream>()
+        {
+          @Override
+          public OutputStream getOutput() throws IOException
+          {
+            return fs.create(path);
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Will this retry on transient errors? It seems like we would want a copy to be able to retry (and overwrite if prior was a partial success) when feasible.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 3 above)
Comment (xvrl) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Most of this code is unchanged, it was just moved into a separate method, so I think it's ok to leave this as is unless it's new code.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 5]
```diff
@@ -109,34 +121,96 @@ public static void setupClasspath(
...
+          }
+        }
+    );
+  }
+
+  private static boolean isSnapshot(File jarFile)
+  {
+    return jarFile.getName()
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: (non-blocker) under the hood this does ```        Pattern p = Pattern.compile(regex);
        Matcher m = p.matcher(input);``` for every jar file. Is it possible to simply have a static `Matcher`, or `Pattern`? Note that `Matcher` instances are not thread safe but `Pattern` is.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 2 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: fixed.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 5 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: fixed.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 4 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: can you elaborate more on which transient error you are referring to ? 
At present, this will overwrite any partial written file from previous run if any.

Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Added Retry logic to JobHelper.setupClasspath

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 6]
```diff
@@ -109,34 +124,150 @@ public static void setupClasspath(
...
+          } else {
+            exception = e;
+          }
+        }
+        if(exception != null){
+          throw exception;
+        }
       }
     }
+    DistributedCache.addFileToClassPath(hdfsPath, conf, fs);
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: DistributedCache is deprecated, can you use `job.addFileToClassPath` to add the files?

Yes, right now that seems to do exactly this under the hood, but if hadoop ever gets their crap together they can change this without warning.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 7]
```diff
@@ -109,34 +124,150 @@ public static void setupClasspath(
...
+  {
+    // Snapshot jars are uploaded to non shared intermediate directory.
+    final Path hdfsPath = new Path(intermediateClassPath, jarFile.getName());
+
+    // existing is used to prevent uploading file multiple times in same run.
+    if (!existing.contains(hdfsPath)) {
+      uploadJar(jarFile, hdfsPath, fs);
+      existing.add(hdfsPath);
+    }
+    DistributedCache.addFileToClassPath(hdfsPath, conf, fs);
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Same about using `job.addFileToClassPath`

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 6 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: done.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 7 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: done.

**Comment by xvrl** [comment]
@nishantmonu51 @drcrallen looks like this change is a little less trivial than originally thought. Should we push it back to 0.8.3 / 0.9.0?

File: indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java [hunk 8]
```diff
@@ -140,11 +140,17 @@ public static Path getTaskPath(JobID jobID, TaskAttemptID taskAttemptID, Path wo
...
+  {
+    return new Path(workingDirectory, jobName.replace(":", ""));
+  }
+
   public static void cleanup(Job job) throws IOException
   {
     final Path jobDir = getJobPath(job.getJobID(), job.getWorkingDirectory());
     final FileSystem fs = jobDir.getFileSystem(job.getConfiguration());
     fs.delete(jobDir, true);
+    fs.delete(getJobClassPathDir(job.getJobName(), job.getWorkingDirectory()), true);
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java]: This method could use some extra assurances / retries as well. (non-blocking)

**Comment by drcrallen** [comment]
@xvrl I'm actually ok with it right now. I would love to have some more improvements on it and Unit tests, which can come in 0.9.x but at this point I think it is in a *much* better place than it was, and I'm :+1: if @nishantmonu51 is willing to add some more assurances and unit tests for 0.9

**Comment by nishantmonu51** [comment]
I tried to write a Unit test for above, but the JobHelper.setupClasspath skips the jar copying if the fileSystem is LocalFileSystem, so i tested it manually with local HDFS setup. 
For adding retry logic for task cleanups, I think we can add that in 0.9, will create an issue for doing that in a separate PR.

File: indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java (hunk 8 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java]: I think we need to do this for cleanup for other Hadoop Jobs as well, would like to do it in a separate PR. Opening a github issue

Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/updater/HadoopConverterJob.java]: https://github.com/druid-io/druid/issues/1828

**Comment by nishantmonu51** [comment]
btw, I am fine with moving this to 0.9/0.8.3 also.

**Comment by drcrallen** [comment]
@nishantmonu51 io.druid.segment.loading.HdfsFileTimestampVersionFinderTest#setupStatic sets up a mini hdfs cluster for actual hdfs testing.

**Comment by nishantmonu51** [comment]
@drcrallen thanks, I will use that and add a Test.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 9]
```diff
@@ -90,9 +93,21 @@ public static Path distributedClassPath(Path base)
     return new Path(base, "classpath");
   }
 
+  /**
+   * Uploads jar files to hdfs and configures the classpath.
+   * Snapshot jar files are uploaded to intermediateClasspath and not shared across multiple jobs.
+   * Non-Snapshot
```
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Non-Snapshot? Is this incomplete sentence?

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 10]
```diff
@@ -109,34 +124,151 @@ public static void setupClasspath(
     if (fs instanceof LocalFileSystem) {
       return;
     }
-
+    // Create distributed directory if it does not exist.
+    fs.mkdirs(distributedClassPath);
```
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: this did not exist before, why is this needed now? and also why not fs.mkdirs(intermediateClassPath) too?

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 9 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: fixed.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 10 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: FileSystem.rename Fails if the parent of dst does not exist or is a file, added it to ensure that the parent dir exists. have added comment in code also.

**Comment by nishantmonu51** [comment]
@drcrallen added test.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 11]
```diff
@@ -111,32 +125,152 @@ public static void setupClasspath(
...
+      @Override
+      public boolean apply(Throwable input)
+      {
+        if (input == null) {
+          return false;
+        }
+        if (input instanceof IOException) {
+          return true;
         }
+        return apply(input.getCause());
```
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: addJarToClassPath has 
```
throw new ISE("File does not exist even after moving from[%s] to [%s]", intermediateHdfsPath, hdfsPath);
```

that means it won't be retried. Is that  intentional?

why not retry here irrespective of what the exception type is, in the worst case there would be 3 retries instead of having to do a recursion?

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 10 above)
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: with the comment it is clearer, thanks.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 11 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: will modify it to IOException though this should ideally never happen in any case.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 12]
```diff
@@ -79,6 +79,8 @@
   private static final int NUM_RETRIES = 8;
   private static final int SECONDS_BETWEEN_RETRIES = 2;
   private static final int DEFAULT_FS_BUFFER_SIZE = 1 << 18; // 256KB
+  private static final Pattern SNAPSHOT_JAR = Pattern.compile(".*SNAPSHOT(-selfcontained)?\\.jar$");
```
Comment (himanshug) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: nit: this should be
```
Pattern.compile(".*\\-SNAPSHOT(-selfcontained)?\\.jar$")
```
just in case someone was crazy enough to have artifact name be something like iAmNotSNAPSHOT.jar , it is not a "snapshot" jar really.

Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: done.

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 11 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: done.

**Comment by himanshug** [comment]
LGTM

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 13]
```diff
@@ -111,32 +125,158 @@ public static void setupClasspath(
...
+        try {
+          if (fs.exists(intermediateHdfsPath)) {
+            fs.delete(intermediateHdfsPath, false);
+          }
+        }
+        catch (IOException e) {
+          if (exception != null) {
+            exception.addSuppressed(e);
+          } else {
+            exception = e;
```
Comment (guobingkun) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Is this just making sure we don't have a jar left over at ```intermediateHdfsPath```?

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+  @AfterClass
+  public static void tearDownStatic() throws IOException
+  {
+    if (miniCluster != null) {
+      miniCluster.shutdown(true);
+    }
+  }
+
+  @After
+  public void tearDown() throws IOException
```
Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: Can you add checks here to make sure the cleanup works as intended. Like make sure the files do not exist after the deletes are called?

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+    miniCluster.getFileSystem().delete(finalClasspath, true);
+    miniCluster.getFileSystem().delete(intermediatePath, true);
+  }
+
+  @Test
+  public void testAddSnapshotJarToClasspath() throws IOException
+  {
+    Job job = Job.getInstance(conf, "test-job");
+    DistributedFileSystem fs = miniCluster.getFileSystem();
+    Path intermediatePath = new Path("/tmp/classpath");
```
Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: Suggest a `@Rule TemporaryFolder` to house this.

Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: oh wait, is this relative to the HDFS cluster?

Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: Intermediate path should be unique per task, right? can this be randomized a bit so one test cannot pollute another?

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class HdfsClasspathSetupTest
+{
+  private static MiniDFSCluster miniCluster;
+  private static File hdfsTmpDir;
+  private static Configuration conf;
+  public static Path finalClasspath = new Path("/tmp/classpath");
+  public static Path intermediatePath = new Path("/tmp/intermediate");
```
Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: Can this be randomized per run so different tests don't collide with each other?

**Comment by drcrallen** [comment]
@nishantmonu51 is there a way to add tests for failure cases?

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 13 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: yeah.

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java (hunk 16 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: this is hdfs path, do we still need to randomize it ? 
I think different runs should get different tempFile for hdfsTmpDir.

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java (hunk 15 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: same here, this is on hdfs, diff runs will have diff hdfs directory.

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java (hunk 16 above)
Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: Ah sure, if that's the case then no problem

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java (hunk 13 above)
Comment (guobingkun) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: I see :)  Can we also invert the condition in "if else"? I actually feel it takes a bit less mind gymnastics to understand if it was 
``` java 
if (exception == null) {
  exception = e;
} else {
  exception.addSuppressed(e);
}

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+          pool.submit(
+              new Callable()
+              {
+                @Override
+                public Boolean call() throws Exception
+                {
+                  int id = barrier.await();
+                  Job job = Job.getInstance(conf, "test-job-" + id);
+                  Path intermediatePathForJob = new Path(intermediatePath, "job-" + id);
+                  barrier.await();
```
Comment (guobingkun) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: not sure why we need to await() for the second time?

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java (hunk 14 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: done.

**Comment by nishantmonu51** [comment]
I added a test for concurrent upload which tries to simulate the failure case which this PR is intended to fix.

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java (hunk 16 above)
Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: The static BeforeClass should only be run once, not per `@Test`. As such this probably needs randomized.

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+                      job.getConfiguration().get(MRJobConfig.CLASSPATH_FILES)
+                  );
+                  return true;
+                }
+              }
+          )
+      );
+    }
+
+    for (Future<Boolean> future : futures) {
```
Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: suggest Futures.allAsList(futures).get(....)

Comment (drcrallen) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: Then you don't really need to check the return value because it should bubble up any exceptions in the execution.

Comment (nishantmonu51) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: nice suggestion, done.

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java (hunk 16 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: the @After cleanup method makes sure that all the contents gets deleted once a test completes, will randomize it too, just in case someone makes any changes there in future.

**Comment by drcrallen** [comment]
:+1:

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+  private File dummyJarFile;
+  private Path finalClasspath;
+  private Path intermediatePath;
+
+  @BeforeClass
+  public static void setupStatic() throws IOException, ClassNotFoundException
+  {
+    hdfsTmpDir = File.createTempFile("hdfsClasspathSetupTest", "dir");
+    hdfsTmpDir.deleteOnExit();
+    if (!hdfsTmpDir.delete()) {
```
Comment (himanshug) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: why is `hdfsTmpDir.deleteOnExit()` needed when `hdfsTmpDir.delete()` is ensured?
Also, it is safer to use @Rule TemporaryFolder instead as that will do cleanup as soon as test is done instead of jvm exit.

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  @Before
+  public void setUp() throws IOException
+  {
+    // intermedatePath and finalClasspath are relative to hdfsTmpDir directory.
+    intermediatePath = new Path(String.format("/tmp/classpath/%s", UUIDUtils.generateUuid()));
+    finalClasspath = new Path(String.format("/tmp/intermediate/%s", UUIDUtils.generateUuid()));
+    dummyJarFile = File.createTempFile("dummy-test", ".jar");
+    dummyJarFile.deleteOnExit();
```
Comment (himanshug) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: can you use TemporaryFolder?

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java (hunk 19 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: I tried using TemporaryFolder, but we cant use it as a static field.

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java (hunk 20 above)
Comment (nishantmonu51) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: done.

File: indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java (hunk 19 above)
Comment (himanshug) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: hmmm, if this is to be done only once for all the tests here then we can probably make things non static and use a no argument constructor to do the setup? Is there a reason for hdfsTmpDir to be static?

Comment (nishantmonu51) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: I can use a non arg constructor, but still i will need to make the fields static since i will have to shut it down in a static afterClass method. I think it will be wierd to initialize static fields of a class in constructor rather than a BeforeClass method.

Comment (himanshug) [inline indexing-hadoop/src/test/java/io/druid/indexer/HdfsClasspathSetupTest.java]: ok, i see, junit unfortunately does not have anything to setup/teardown "instance" .
